### ISSUE TITLE
Duplicated chord diagrams from {define: }

### ISSUE THREAD

**Issue by gwyndaf** [ChordPro/chordpro#282]
with v6.000_025
```
{title: Custom chord diagram duplication}
{define: Am/C copy Am}
{define: C6 copy Am}

Slash chord [Am/C] and root position chord [Am]
Alternate chord [C6]
```
Processing the above with my own configs puts three identical Am chord diagrams on the page, corresponding to the original Am chord and the two `{define: copy}` directives.

Processing with `-X` generates three identical _diagrams_, but with different names. I expect that's because there's no `display` value in the default instrument config, whereas my own include `display` settings.

It's not a big issue, but duplicating the same diagram and display name puts unnecessary extra information on the page, and I try to keep it lean (mainly to avoid confusing learners).

My reason for defining alternative chords (not simply replacing in source) is to simplify for some instruments/players only, while retaining the more complex chord forms for other instruments/players.

Is there a way of avoiding such duplication? (Probably an enhancement)

**Comment by sciurius** [comment]
If I understand correctly: You use three chords in your song, Am/C, Am and C6. You get three chord diagrams, Am/C, Am and C6. It doesn't matter that these diagrams are the same. Also, which of the three would be eligible for hiding? If you leave out e.g. Am, an attentive learner would be confused -- hey, where is Am?

So it is not clear to me what enhancement you're looking for.
You can suppress chord diagrams by putting them in the `diagrams.suppress` list (see https://www.chordpro.org/beta/chordpro-configuration-generic/#printing-chord-diagrams).

**Comment by gwyndaf** [comment]
Not exactly. If two chords are redefined with `copy Am` (e.g. as simpler alternative), three Am chord diagrams get output, all visually identical, including the name:
[Test.pdf](https://github.com/ChordPro/chordpro/files/11642428/Test.pdf)

Different chord names with the same voicing/fingering (e.g. Am7 and C6 on ukulele) should still be output.

Thanks for the `diagrams.suppress` suggestion, which does seem to work at song level:
```
{define: Am/C copy Am}
{define: C6 copy Am}
{+diagrams.suppress Am/C }
{+diagrams.suppress C6 }
```
successfully removes the duplicate diagrams, _but_ it's universal. Many of my chord simplifications are conditional on instrument, e.g. for ukulele players, while guitarists get full slash chords etc. `suppress` seems to be based on the source/input name rather than the output name. So it would also suppress guitar output of Am/C where it's a different diagram from Am.

Perhaps it's too complicated (or subjective) to suppress duplicate output diagrams automatically, so maybe `{define: }` could take an additional `diagram` parameter to control diagram output on a per-chord/per-instrument basis? Maybe just a `true`/`false` value to meet the immediate need.

That approach might have future potential for something like `{define: Am/C copy Am diagram "<span foreground=blue>%{diag_grid}</span>" }` as a way to visually differentiate chord diagrams (and on a per-instrument basis), e.g. principal or optional/passing chords. A separate parameter would allow diagram output to be controlled without affecting how chord names get displayed in the song body. But I don't  know if applying markup to the diagram grid is even feasible.

'Learners' was probably a poor choice of word, because that suggests taking time to study song, in a taught session. 'Beginners' at a playing session is more accurate: they'll often glance at the number of chord diagrams, and consider a song difficult if there are too many, without taking time to notice if some are duplicates! Certainly not a high priority, just a way of making the user (i.e. reader/player) experience clear and simple :)

**Comment by sciurius** [comment]
What input did you use for Test.pdf?

> successfully removes the duplicate diagrams, but it's universal. Many of my chord simplifications are conditional on instrument, e.g. for ukulele players, while guitarists get full slash chords etc. 

Do you need something as `{+diagrams.suppress-guitar C6}` ? I can make that work.

> maybe {define: } could take an additional diagram parameter

> diagram "&lt;span foreground=blue>%{diag_grid}&lt;/span>"

No kidding, this could be possible in a future release...

In the mean time, would you settle for a `define` argument `diagram` with values `none`, `default` or a colour?

**Comment by gwyndaf** [comment]
Original input (before adding `suppress` lines) was as above:
```
{title: Custom chord diagram duplication}
{define: Am/C copy Am}
{define: C6 copy Am}

Slash chord [Am/C] and root position chord [Am]
Alternate chord [C6]
```

A `diagram` parameter for `{define: }` would be great, and exploit existing conditional functionality (for instrument-specific variants). My current need is purely whether or not to display them (`none`/`default`). Colour (e.g. grey for optional/passing chords) or markup was just where my imagination took me from there, but seems more a possibility than a problem to solve! :)

It's probably unnecessary duplication to also make `diagrams.suppress` conditional, if an enhanced `{define: diagram }` directive will do the job.

From a usage point of view, it seems more elegant to deal with all the chord-specific instructions in a single `{define: }` directive (not an additional `suppress` line). I can't see myself needing to hide any default chords from the config file, so all custom (and potentially duplicated) chords would already have a `{define}` directive in the song, and that seems a better place at which to hide them.

**Comment by sciurius** [comment]
> Original input (before adding suppress lines) was as above:

I get different results, see the attached result, which is imo correct.
[x.pdf](https://github.com/ChordPro/chordpro/files/11643301/x.pdf)

Also, I agree that tricking `diagrams.suppress` is no longer relevant once `define` has a `diagram` argument.

**Comment by gwyndaf** [comment]
Sorry, probably my mistake. I tried it again with `-X`and get the same result as you (same voicings, different names), which does seem reasonable.

Maybe I ran it before with `-X` plus my own guitar config, which includes `display` values, and I guess they supersede the default chord names from the `{define: }` directives, i.e. explicit display values for "Am" get copied from the config as the display name for all three chords, resulting in three identical diagrams and names.
